students/TrkCat/Lesson05/mailroom.py

```diff
@@ -0,0 +1,154 @@
+    if donor_name.lower() not in donors:
+        donors[donor_name.lower()] = {'name': donor_name}
+    cur_dict = donors[donor_name.lower()]
+    while True:
+        try:
+            donation = float(input('Enter the donation amount: '))
+            break
+        except ValueError:
+            print('Error: Please enter a numeric value for the dollar amount.')
+    cur_dict.setdefault('donations', []).append(donation)
+    cur_dict['last_don'] = donation
+    update_tot_avg(cur_dict)
+    letter = gen_email()
+    print(letter.format(**cur_dict))
+    return False
+
+
+def list_donors():
+    """Print list of current donors"""
+    print('\n\nCurrent Donors:\n')
+    for donor in donors.values():
+        print(donor['name'])
+
+
+def gen_email():
+    """Return a thank you email"""
+    letter = ('\nFROM: Your friendly local charity mailroom.\n'
+              'TO: {name}\n'
+              'RE: Your recent donation\n\n'
+              '\nThank you so much for your recent donation of'
+              ' ${last_don:,.2f}. This will go a long way towards helping to'
+              ' save the pythons. Your generosity is most appreciated!'
+              '\n\nBest Regards,\nSave The Pythons\n'
+              )
+    return letter
+
+
+def quit_menu():
+    """Quit current menu"""
+    return False
+
+
+def update_tot_avg(don_dict):
+    """Update total and average donation value in dict d."""
+    don_dict['tot_don'] = sum(don_dict['donations'])
+    don_dict['num_don'] = len(don_dict['donations'])
+    don_dict['avg_don'] = don_dict['tot_don']/don_dict['num_don']
+
+
+def show_menu(prompt, disp_dict):
+    """Generate menu with dispatch dictionary"""
+    while True:
+        sel = input(prompt)
+        try:
+            if disp_dict[sel]() is False:
+                return
+        except KeyError:
+            print('Error: Please enter an integer from the menu only.')
+
+
+if __name__ == '__main__':
+    donors = {'bill gates': {'name': 'Bill Gates',
+                             'donations': [789.25, 87562.22, 125000.00],
+                             'last_don': 125000
+                             },
+              'jeff bezos': {'name': 'Jeff Bezos',
+                             'donations': [3456.89, 130],
+                             'last_don': 130
+                             },
+              'jimmy buffett': {'name': 'Jimmy Buffett',
+                                'donations': [85000],
+                                'last_don': 85000
+                                },
+              'abe lincoln': {'name': 'Abe Lincoln',
+                              'donations': [5, 2, 1],
+                              'last_don': 1
+                              },
+              'yankee doodle': {'name': 'Yankee Doodle',
+                                'donations': [67],
+                                'last_don': 67
+                                }
+              }
+    [update_tot_avg(donor) for donor in donors.values()]
```

Comment:
Thanks Rob!

I looked through the code and this is the only place I could find to use comprehension. Is there a better/more appropriate way to do this? I previously just had an if statement. I wasn't trying to generate a new dict, just populate the values in the initial one.

Reply:
In this case, a simple for loop might be the better choice and I'm not seeing much else as far as obvious opportunities to implement comprehensions in your code. That's been the case for quite a few students with this assignment, and it will totally depend on your mailroom implementation as to whether or not comprehensions will be useful. 

You might look at the `create_report` function, since we'll be adding unit testing in the next lesson, you'll be looking for ways to refactor your functions so that they can have testable (and printable) return values, rather than printing as a 'side effect' of the function. 

A comprehension might be useful there for building the report as a list of strings, and then joining them together as your return value. So maybe think about that if you're looking for places to refactor and make things more testable in the next lesson.

Reply:
Thanks, I'll definitely look into it.